### PR TITLE
Adding Header.Select fromSafe, toRaw multiple, and Aux

### PR DESCRIPTION
**Pull request by zarthross**
Some enhancements needed to help enable Rho support. 

See: https://github.com/http4s/rho/pull/501 for some usage example.

I could use feedback on these, once I get buy in I'll happily add/fixup the unit tests!

Resolves https://github.com/http4s/http4s/issues/4706

### CODE REVIEW

**Comment by rossabaker** [comment]
We should probably target 0.22 with this: that's where the new header model debuts.  It shouldn't be hard to merge to main, since it's pure work.

File: core/src/main/scala/org/http4s/Header.scala
Comment (rossabaker) [inline core/src/main/scala/org/http4s/Header.scala:154]: Trying to remember why we have an `Option[F]` on these, resulting in an `Option[NonEmptyList]` for `Recurring` instead of `F` being `Option` or `List`.

I don't love the name `fromSafe`, but I do think the `Ior` is a correct model: "give me all you can, and warn me about the ones you couldn't."

File: core/src/main/scala/org/http4s/Header.scala
Comment (rossabaker) [inline core/src/main/scala/org/http4s/Header.scala:148]: I'm a bit lost on this one.  Is this something rho needed?  Can you point me to it?

File: core/src/main/scala/org/http4s/Header.scala
Comment (zarthross) [reply core/src/main/scala/org/http4s/Header.scala:154]: > Trying to remember why we have an Option[F] on these, resulting in an Option[NonEmptyList] for Recurring instead of F being Option or List.

I could look into changing that, it makes more sense to me.

> I don't love the name fromSafe, but I do think the Ior is a correct model: "give me all you can, and warn me about the ones you couldn't."

I think just having 1 `from` that returns the `IOr` would be fine, and our users can `toOption` if they want to ignore the parse failures.

File: core/src/main/scala/org/http4s/Header.scala
Comment (zarthross) [reply core/src/main/scala/org/http4s/Header.scala:148]: https://github.com/http4s/rho/blob/c5a90fcebaee2524d5898969f38ec183b2eb8c86/core/src/main/scala/org/http4s/rho/RhoDslHeaderExtractors.scala#L146

I debated about just changing the existing `toRaw` rather than having a separate `toRawOne`, since most of the time you have a header with a semigroup, rather than a list of something, (or at least that seems to be the http4s pattern) where you might want to just do 1 of.

**Comment by zarthross** [comment]
> We should probably target 0.22 with this: that's where the new header model debuts. It shouldn't be hard to merge to main, since it's pure work.

Done.

File: core/src/main/scala/org/http4s/Header.scala
Comment (zarthross) [reply core/src/main/scala/org/http4s/Header.scala:148]: Have any suggestions on this @rossabaker ?  Would it be better to have just `toRaw` and not a `toRawOne`?

File: core/src/main/scala/org/http4s/Header.scala
Comment (zarthross) [reply core/src/main/scala/org/http4s/Header.scala:154]: @rossabaker thoughts?

**Comment by rossabaker** [comment]
Haven't forgotten this.  0.21 sucked up all my time, but this is absolutely necessary before 0.22 so we don't leave rho hanging.  Extra opinions welcome.

File: core/src/main/scala/org/http4s/Header.scala
Comment (aeons) [reply core/src/main/scala/org/http4s/Header.scala:154]: > I think just having 1 from that returns the IOr would be fine, and our users can toOption if they want to ignore the parse failures.
I think this makes sense.

Comment (rossabaker) [reply core/src/main/scala/org/http4s/Header.scala:154]: > I think just having 1 from that returns the IOr would be fine, and our users can toOption if they want to ignore the parse failures.

I agree with this.

**Comment by rossabaker** [comment]
So this is still in draft.  We want to release 0.22.0, but are blocked by a couple issues and circe.  I also want to unblock Rho.  We could merge this as is and clean it up before the final release, or we could let rho work off snapshots/source dependencies and clean up before the final release.  Thoughts?

**Comment by zarthross** [comment]
I think I can make some time this weekend for the suggestions here, and hopefully any associated testing.

**Comment by rossabaker** [comment]
That sounds great.  I have a stupid busy weekend, so if we can get things greened up, I may do a milestone today.  But I will commit to do another release as soon as this is ready, because we don't want to leave Rho hanging.

File: core/src/main/scala/org/http4s/Header.scala
Comment (zarthross) [reply core/src/main/scala/org/http4s/Header.scala:154]: What about `Headers.get`? Should it return an `Option` or should it also return an `Option[Ior...`?

File: core/src/main/scala/org/http4s/Header.scala
Comment (zarthross) [reply core/src/main/scala/org/http4s/Header.scala:148]: @rossabaker ^?

**Comment by zarthross** [comment]
Just 2 questions that need answering on the above issues, and I'd be ready to move this out of Draft.

File: core/src/main/scala/org/http4s/Header.scala
Comment (rossabaker) [reply core/src/main/scala/org/http4s/Header.scala:153]: `toRawOne` is equivalent to `toRaw` for single headers, and for headers with a Semigroup instance?  Only recurring headers without a Semigroup need this new signature?  And that is in practice only `Set-Cookie`?

Comment (rossabaker) [reply core/src/main/scala/org/http4s/Header.scala:153]: I don't think `toRaw(a: F[A]): Header.Raw` is sensible for `Set-Cookie`, because a `NonEmptyList` of `Set-Cookie` can't be reduced to a single header.

Comment (zarthross) [reply core/src/main/scala/org/http4s/Header.scala:153]: What would be correct. I'm fine with dropping the `toRawOne`, but Rho will need the new version of `toRaw` that takes a `F[A]` instead of just an `A` in order for the types to work out.

File: core/src/main/scala/org/http4s/Header.scala
Comment (rossabaker) [reply core/src/main/scala/org/http4s/Header.scala]: Hmm.  Good question.  `My inclination is `Option[Ior[...`, but that would be a massive breaking change.  The existing signature is consistent with its map-like semantics.

I argued against the overload earlier, and I don't think the typeclass needs the overload, but `Headers` is the primary user API.  Maybe leaving `get` as is, and `getWithWarnings` that returns the `Option[Ior[...`?

File: core/src/main/scala/org/http4s/Header.scala
Comment (rossabaker) [reply core/src/main/scala/org/http4s/Header.scala:153]: I guess what I'm saying is that there's no correct implementation of that signature for `Set-Cookie`, because you can't just concatenate the values of the `NonEmptyList` with a comma.  The return type would need to be a collection for that header.  It's the exception to all the other rules.

Comment (zarthross) [reply core/src/main/scala/org/http4s/Header.scala:153]: Sooo.... are you saying we keep both signatures?  Do we just need better names?

File: core/src/main/scala/org/http4s/Header.scala
Comment (zarthross) [reply core/src/main/scala/org/http4s/Header.scala]: Done.

File: core/src/main/scala/org/http4s/Header.scala
Comment (rossabaker) [reply core/src/main/scala/org/http4s/Header.scala:153]: I'm concerned about what the need for this signature is telling us, because the one that joins with a comma is an invalid implementation.  Perhaps I need to spend more time studying that Rho link.

Comment (rossabaker) [reply core/src/main/scala/org/http4s/Header.scala:153]: `cats-effect-3` branch, right?  I might need to get hands on with this.  The types are more than I can contemplate on GitHub.

Comment (zarthross) [reply core/src/main/scala/org/http4s/Header.scala:153]: Correct.  The important part for that piece of code is we have an `F[A]` where `toRaw` requires and `A`.   I can probably work around this, but it just seemed logical at the time that `toRaw` should be able to convert to and from raw in the `F[A]` form.

Comment (rossabaker) [reply core/src/main/scala/org/http4s/Header.scala:153]: Yeah, it's just that one exceptional header that makes that not work.  I'm playing type tetris with a local checkout now.

Comment (rossabaker) [reply core/src/main/scala/org/http4s/Header.scala:153]: Okay, https://github.com/http4s/rho/pull/538 compiles after my PR to this PR.  I don't know rho nearly as well as you, so I don't know whether it's acceptable, but maybe that's our way out of this?

Comment (zarthross) [reply core/src/main/scala/org/http4s/Header.scala]: I think it'll be fine.  Thanks for looking into this!

File: core/src/main/scala/org/http4s/syntax/HeaderSyntax.scala
Comment (rossabaker) [inline core/src/main/scala/org/http4s/syntax/HeaderSyntax.scala:42]: This seemed shady to me, but would have changed a lot to rename it to `toRaw1` where it's used.  I'm also not comfortable with how it interacts with `SelectOpsMultiple`.  We might want to take a look at this, still.

**Review by rossabaker — DISMISSED** [review]
I think this is sufficient, and we can do an RC with it when circe-0.14 is ready and make sure Rho has a path forward.

File: core/src/main/scala/org/http4s/syntax/HeaderSyntax.scala
Comment (zarthross) [reply core/src/main/scala/org/http4s/syntax/HeaderSyntax.scala:42]: Yeah... there could be instances of ambiguous implicits...  maybe we can rename `SelectOpsMultiple.toRaw` to something else to avoid that?

Comment (rossabaker) [reply core/src/main/scala/org/http4s/syntax/HeaderSyntax.scala:42]: I think the syntax needs to be consistent with the typeclass: https://github.com/zarthross/http4s/pull/2

**Comment by rossabaker** [comment]
Looks like we also broke this:

```
 ==> X org.http4s.client.middleware.LoggerSuite.RequestLogger should not affect a Post  0.008s org.http4s.ParseFailure: Invalid Content-Type header: Content-Type header doesn't support media ranges
```

**Comment by zarthross** [comment]
> Looks like we also broke this:
> 
> ```
>  ==> X org.http4s.client.middleware.LoggerSuite.RequestLogger should not affect a Post  0.008s org.http4s.ParseFailure: Invalid Content-Type header: Content-Type header doesn't support media ranges
> ```

I can take a look.

**Pull request merged**
